src/Zigzag_persistence/include/gudhi/filtered_zigzag_persistence.h

```diff
@@ -0,0 +1,491 @@
+   * Assumed to be always larger or equal to previously used filtration values or always smaller or equal than previous
+   * values, ie. the changes are monotonous.
+   * @return Number of the operation.
+   */
+  internal_key remove_face(face_key faceID, dimension_type dimension, filtration_value filtrationValue) {
+    ++numArrow_;
+
+    if (dimMax_ != -1 && dimension > dimMax_) {
+      pers_.apply_identity();
+      return numArrow_;
+    }
+
+    auto it = handleToKey_.find(faceID);
+    GUDHI_CHECK(it != handleToKey_.end(), "Zigzag_persistence::remove_face - face not in the complex");
+
+    if (filtrationValue != previousFiltrationValue_)  // check whether the filt value has changed
+    {  // consecutive pairs (i,f), (j,f') mean faces of index k in [i,j-1] have
+      previousFiltrationValue_ = filtrationValue;  // filtration value f
+      filtrationValues_.emplace_back(numArrow_, previousFiltrationValue_);
+    }
+
+    pers_.remove_face(it->second, dimension);
+    handleToKey_.erase(it);
+
+    return numArrow_;
+  }
+
+  /**
+   * @brief To use when a face is neither inserted nor removed, but the filtration moves along the identity operator
+   * on homology level. Useful to keep the birth/death indices aligned when insertions/removals are purposely skipped
+   * to avoid useless computation.
+   * @return Number of the operation.
+   */
+  internal_key apply_identity() {
+    ++numArrow_;
+    pers_.apply_identity();
+    return numArrow_;
+  }
+
+  /**
+   * @brief Returns the "index persistence diagram" of the current filtration, that is, the pairs of atomic arrow
+   * numbers corresponding to a birth-death pair. Does not contain points at infinity, only the cycle classes which
+   * already died are represented.
+   *
+   * @return Reference to the list of intervals.
+   */
+  const std::vector<Index_interval>& get_index_persistence_diagram() const { return persistenceDiagram_; }
+
+  /**
+   * @brief Returns the filtration values \f$[f(b),f(d)]\f$ associated to the indices \f$[b,d]\f$ which are retrieved
+   * by @ref get_index_persistence_diagram.
+   *
+   * @param birthKey Birth index
+   * @param deathKey Death index
+   * @return A pair of filtration values associated to the given indices.
+   */
+  std::pair<filtration_value, filtration_value> map_index_to_filtration_value(internal_key birthKey,
+                                                                              internal_key deathKey) const {
+    // filtration_values_ must be sorted by increasing keys.
+    auto itBirth =  // lower_bound(x) returns leftmost y s.t. x <= y
+        std::lower_bound(
+            filtrationValues_.begin(), filtrationValues_.end(),
+            birthKey,
+            [](std::pair<internal_key, filtration_value> p, internal_key k) {
+              return p.first < k;
+            });
+    if (itBirth == filtrationValues_.end() || itBirth->first > birthKey) {
+      --itBirth;
+    }
```

Comment:
We do this at least 3 times, it would be nice to split it into a function (it could even have the same name). Actually, once we have a function to translate one index, I am not sure the one that translates a pair is that useful, especially since it is used only once.